File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow, fast = head, head
        
        # traverse to just before middle point
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        # cut off to split LL's
        head2 = slow.next
        slow.next = None
        
        # reverse second half in place
        cur = head2
        prev = None
        while cur:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        
        head2 = prev
        
        # re order with first and second half
        cur = head
        cur2 = head2

        while cur2:
            next, next2 = cur.next, cur2.next
            cur.next = cur2
            cur2.next = next
            cur = next
            cur2 = next2

File: Data Structures & Algorithms/decode-ways/submission-1.cpp
class Solution {
/*
10-1-2
J A B
10-12
J L

s = 1012
arr:
1 0 0 0 0
1 1 1 1 2
if i != 0 && i and i-1 valid then dp[i] = dp[i-1] + dp[i-2]
else dp[i] = dp[i-1]

s = 06
arr:
1 0 0
1 0 0
1 0 1

s = 12345
arr:
1 0 0 0 0 0
1 1 2 3 3 3

*/
public:
    int numDecodings(string s) {
        int n = s.length();
        int cur = 1, prev = 0;

        for (int i = 0; i < n; ++i) {
            int next = 0;
            if (s[i] - '0' != 0) {
                next += cur;
            }
            if (i > 0 && (s[i-1] - '0' == 1) || (s[i-1] - '0' == 2 && s[i] - '0' <= 6)) {
                next += prev;
            }

            prev = cur; cur = next;
        }

        return cur;
    }
};
